common/network-common/src/main/java/org/apache/spark/network/protocol/Encoders.java

```diff
@@ -44,6 +51,71 @@ public static String decode(ByteBuf buf) {
     }
   }
 
+  /** Bitmaps are encoded with their serialization length followed by the serialization bytes. */
+  public static class Bitmaps {
+    public static int encodedLength(RoaringBitmap b) {
+      // Compress the bitmap before serializing it
+      b.trim();
+      b.runOptimize();
+      return 4 + b.serializedSizeInBytes();
+    }
+
+    public static void encode(ByteBuf buf, RoaringBitmap b) {
+      ByteBuffer outBuffer = ByteBuffer.allocate(b.serializedSizeInBytes());
+      try {
+        b.serialize(new DataOutputStream(new OutputStream() {
```

Comment:
Good point, I think this also avoids one more memory copy.

Reply:
We will still have a memory copy, right ? serialize to local ByteBuffer, and copy from local ByteBuffer to Bytebuf

Reply:
We can get the underlying ByteBuffer for the given ByteBuffer, serialize from/deserialize into it, and then update the write/read indexes accordingly.
This is reflected in the latest change.

Reply:
Also, realized that the reason we were not using this API initially is because of the old version of RoaringBitmap in Spark 2.3. This API was not available there.